subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/PerformanceReportScenario.groovy

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.performance.results
+
+/**
+ * Represents a row in performance report, i.e. a specific scenario.
+ */
+class PerformanceReportScenario {
+    /**
+     * The executions read from TeamCity-build-generated-result-JSONs.
+     */
+    final List<PerformanceReportScenarioTeamCityExecution> teamCityExecutions
+
+    /**
+     * The execution read from performance database which has the same TC build id as `teamCityExecutions`.
+     */
+    final List<PerformanceReportScenarioHistoryExecution> currentExecutions
+
+    /**
+     * The execution read from performance database, excluding current executions
+     */
+    final List<PerformanceReportScenarioHistoryExecution> recentExecutions
+
+    final boolean crossBuild
+
+    final boolean fromCache
+
+    PerformanceReportScenario(
+            List<PerformanceReportScenarioTeamCityExecution> teamCityExecutions,
+            List<PerformanceReportScenarioHistoryExecution> historyExecutions,
+            boolean crossBuild,
+            boolean fromCache
+    ) {
+        if (teamCityExecutions.empty) {
+            throw new IllegalArgumentException("teamCity executions must not be empty!")
+        }
+        this.teamCityExecutions = teamCityExecutions
+        this.crossBuild = crossBuild
+        this.fromCache = fromCache
+
+        Set<String> teamCityBuildIds = teamCityExecutions.collect { it.teamCityBuildId }.toSet()
+        this.currentExecutions = historyExecutions.findAll {
+            teamCityBuildIds.contains(it.teamCityBuildId)
```

Comment:
For CI this is good, for local builds we should probably use something different since the teamcity build ids will be `null` (or empty strings?). Or do we have sets containgin `null`s here and we check for that?

Reply:
For local builds, this is actually `PerformanceTest.buildId` passed via https://github.com/gradle/gradle/blob/540bf01f15879c211be47889403141182c40ee7d/build-logic/performance-testing/src/main/groovy/gradlebuild/performance/tasks/PerformanceTest.groovy#L174

which is injected by `BUILD_ID` env variable here: https://github.com/gradle/gradle/blob/8fb70f25ac8d55dda13fed0098d903dd0cc12a5e/build-logic/performance-testing/src/main/kotlin/gradlebuild/performance/PerformanceTestPlugin.kt#L406

I added a dummy `localBuild` to make it work.

Reply:
I don't think `localBuild` will work. That would mean that when querying form the database, we would detect all local executions for a scenario as current executions, since the teamcityBuildId is the same. I would need to think some more what we can do here. We should probably fix the local performance report generation in a separate PR, so we can go forward with this one. Or we disable the performance report for local builds, given that you have the Gradle profiler report as well.

Reply:
> That would mean that when querying form the database, we would detect all local executions for a scenario as current executions, since the teamcityBuildId is the same.

Yes, but we have a `limit 5` in the query so it's fine - another solution is to add a timestamp like `localBuild-202106081200` to uniquely identify that local build, WDYT?